### PR TITLE
made va_list methods public

### PR DESCRIPTION
**Pull request by jpsim**
to allow other variadic methods to forward their arguments and to mirror what's done in Foundation APIs.

This addresses #2934. I've marked the methods as `:nodoc:` so they won't be included in our HTML documentation because I don't think they should be.

### CODE REVIEW

**Comment by Rusik** [comment]
Hi @jpsim!
Maybe you should move `RLM_VARARG(predicateFormat, args)` with check
```
if (PREDICATE_FORMAT && ![PREDICATE_FORMAT isKindOfClass:[NSString class]]) {
    NSString *reason = @"predicate must be an NSString with optional format va_list";
    [NSException exceptionWithName:RLMExceptionName reason:reason userInfo:nil];
}
```
 from `+ (RLMResults *)objectsWhere:(NSString *)predicateFormat, ...` to `+ (RLMResults *)objectsWhere:(NSString *)predicateFormat args:(va_list)args`?

**Comment by bdash** [comment]
I agree that pushing the checking performed by `RLM_VARARG` into the `va_list` version would be a good idea if these are made public.

**Comment by jpsim** [comment]
@bdash @Rusik I've moved the predicate format validation into an `RLMValidatedPredicate` function so it's now also applied in the `va_list` methods. Anything else?

File: Realm/RLMQueryUtil.mm [hunk 1]
```diff
@@ -71,6 +71,14 @@ static void RLMPrecondition(bool condition, NSString *name, NSString *format, ..
     return prop;
 }
 
+NSPredicate *RLMValidatedPredicate(id predicateFormat, va_list args) {
+    if (predicateFormat && ![predicateFormat isKindOfClass:[NSString class]]) {
```
Comment (tgoyne) [inline Realm/RLMQueryUtil.mm]: What's the point of this check? The methods are all declared as taking a NSString and we don't check for violations of the static type anywhere else in the api.

Comment (jpsim) [inline Realm/RLMQueryUtil.mm]: That's a *very* good point. Didn't we *used* to have overloads that could be either `NSString *` or `NSPredicate`? I think this is just a leftover from that. I'll revert that back.

File: Realm/RLMQueryUtil.hpp [hunk 2]
```diff
@@ -46,8 +47,7 @@ realm::SortOrder RLMSortOrderFromDescriptors(RLMObjectSchema *objectSchema, NSAr
 // This macro validates predicate format with optional arguments
 #define RLM_VARARG(PREDICATE_FORMAT, ARGS) \
 va_start(ARGS, PREDICATE_FORMAT);          \
-va_end(ARGS);                              \
-if (PREDICATE_FORMAT && ![PREDICATE_FORMAT isKindOfClass:[NSString class]]) {         \
-    NSString *reason = @"predicate must be an NSString with optional format va_list"; \
-    [NSException exceptionWithName:RLMExceptionName reason:reason userInfo:nil];       \
-}
+va_end(ARGS);
```
Comment (bdash) [inline Realm/RLMQueryUtil.hpp]: It looks like the behavior of this macro is wrong (and was wrong prior to this change). The expected usage of `va_list` to forward variadic arguments is:

```c++
va_list args;
va_start(predicate, args);
target_function(predicate, args);
va_end(args);
```

`va_end` is documented as invalidating the `va_list`, so our practice of `va_start` / `va_end`/ `target_function` is relying on undefined behavior.

File: Realm/RLMQueryUtil.hpp [hunk 3]
```diff
@@ -46,8 +47,7 @@ realm::SortOrder RLMSortOrderFromDescriptors(RLMObjectSchema *objectSchema, NSAr
...
+
+// return predicate - throw for invalid format
+NSPredicate *RLMValidatedPredicate(id predicateFormat, va_list args);
```
Comment (bdash) [inline Realm/RLMQueryUtil.hpp]: Why is `predicateFormat` declared as `id` rather than `NSString *`?

File: Realm/RLMQueryUtil.mm (hunk 1 above)
Comment (bdash) [inline Realm/RLMQueryUtil.mm]: Why does this intentionally avoid throwing when passed `nil`?

File: Realm/RLMQueryUtil.mm [hunk 4]
```diff
@@ -71,6 +71,14 @@ static void RLMPrecondition(bool condition, NSString *name, NSString *format, ..
...
+        NSString *reason = @"predicate must be an NSString with optional format va_list";
+        @throw [NSException exceptionWithName:RLMExceptionName reason:reason userInfo:nil];
```
Comment (bdash) [inline Realm/RLMQueryUtil.mm]: Using `@throw RLMException(…)` would tidy this up a little.

File: Realm/RLMQueryUtil.mm (hunk 1 above)
Comment (jpsim) [inline Realm/RLMQueryUtil.mm]: Again, if memory serves me right I think this was left over from very early code where you could pass in `nil` as a filter, which was the equivalent of `allObjects`. Not useful anymore obviously.

**Comment by jpsim** [comment]
Feedback addressed.

**Comment by bdash** [comment]
:+1:

**Pull request merged**